### PR TITLE
Fix NRE when MatchEvaluator returns null

### PR DESCRIPTION
**Pull request by danmosemsft**
Fix https://github.com/dotnet/corefx/issues/39746

This regressed when moving from StringBuilder to ValueStringBuilder as the latter does not allow Append (or Insert) of null, whereas StringBuilder just ignores it.

No fix was needed for RightToLeft, as that uses ValueStringBuilder.AppendReversed, whose parameter is typed ROS<char>, and a cast from a null string to ROS<char> just results in a zero length span. ValueStringBuilder.Append has a string overload apparently for performance reasons which is why it needs null protection.

As far as I can tell all other use in Regex of VSB is passing in strings or chars that are not null, or passing in Spans.

### CODE REVIEW

**Comment by danmosemsft** [comment]
needs 3.0 port

File: src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs [hunk 1]
```diff
@@ -224,6 +224,31 @@ public void Replace_NoMatch()
...
+                        return "x";
+                    case "b":
+                        return null;
+                    case "c":
+                        return "";
+                    case "d":
+                        return "y";
+                    default:
+                        throw new InvalidOperationException();
+                }
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: If you wanted to be fancy with C# 8, you could use a switch expression :)
```C#
(Match match) => match.Value switch
{
    "a" => "x",
    "b" => null,
    "c" => "",
    "d" => "y",
    _ => throw new InvalidOperationException()
}
```

Comment (danmosemsft) [reply src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: Ooo, I'll definitely do that!

File: src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs [hunk 2]
```diff
@@ -224,6 +224,31 @@ public void Replace_NoMatch()
             Assert.Same(input, Regex.Replace(input, "no-match", new MatchEvaluator(MatchEvaluator1)));
         }
 
+        [Theory]
+        [InlineData(RegexOptions.None)]
+        [InlineData(RegexOptions.RightToLeft)]
+        public void Replace_MatchEvaluatorReturnsNullOrEmpty(RegexOptions options)
+        {
+            string input = "abcde";
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/Regex.Replace.Tests.cs]: Not sure why input is defined here in a local but the pattern is defined inline in the Replace call... seems like you could just move "abcde" to be the arg

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Replace.cs [hunk 3]
```diff
@@ -186,7 +186,7 @@ private static string Replace(MatchEvaluator evaluator, Regex regex, string inpu
                             vsb.Append(input.AsSpan(prevat, match.Index - prevat));
 
                         prevat = match.Index + match.Length;
-                        vsb.Append(evaluator(match));
+                        vsb.Append(evaluator(match) ?? "");
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Replace.cs]: You reviewed the other Append call sites and this was the only one that needed the null check?

Comment (danmosemsft) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Replace.cs]: Yes. Most are obvious (char or Span). A couple are not mainly
```
                switch (child.Type())
                {
                    case RegexNode.Multi:
                        vsb.Append(child.Str);
```
After some stepping I don't think I can find a way for `Str` to be null as it originates from succesfully parsing the replacement string. I suppose I could put in a defensive `?? ""` but I think it's OK.

Comment (ZOXEXIVO) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Replace.cs]: It's OK with idle calling of ValueStringBuilder.Append, that have no cases for empty strings ?

Comment (stephentoub) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Replace.cs]: > It's OK with idle call of ValueStringBuilder.Append, that have no cases for empty strings ?

@ZOXEXIVO, I don't understand your question. Can you rephrase?

Comment (ZOXEXIVO) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Replace.cs]: > @ZOXEXIVO, I don't understand your question. Can you rephrase?
Why we call Append with empty string if evaluator(match) return null ?

Comment (stephentoub) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Replace.cs]: This could have instead been:
```C#
string result = evaluator(match);
if (result != null)
{
    vsb.Append(result);
}
```
My guess is @danmosemsft felt it wasn't worth the extra verbosity because nulls are very rare and don't show up on hot paths.  But I'm ok either way if he wants to change it.

Comment (danmosemsft) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Replace.cs]: Right, but I'm happy to change it..
